node-metrics/src/api/node_validator/v0/create_node_validator_api.rs

```diff
@@ -0,0 +1,472 @@
+/// [HotShotEventProcessingTask] is a task that is capable of processing events
+/// that are coming in from a HotShot event stream.  This task will keep an
+/// eye out for ExternalMessageReceived events that can be decoded as a
+/// RollCallResponse.  When a RollCallResponse is received, the public API URL
+/// of the node that sent the message will be sent to the provided url_sender.
+///
+/// Additionally, this can receive Decide events and send the discovered leaves
+/// to the provided leaf_sender.  This can can be used as a means of receiving
+/// leaves that doesn't involve hitting an external service like the task
+/// [ProcessProduceLeafStreamTask] does.
+pub struct HotShotEventProcessingTask {
+    pub task_handle: Option<JoinHandle<()>>,
+}
+
+impl HotShotEventProcessingTask {
+    /// [new] creates a new [HotShotEventProcessingTask] that will process
+    /// events from the provided event_stream.
+    ///
+    /// Calls to [new] will spawn a new task that will start processing
+    /// immediately. The task handle will be stored in the returned structure.
+    pub fn new<S, K1, K2>(event_stream: S, url_sender: K1, leaf_sender: K2) -> Self
+    where
+        S: Stream<Item = Event<SeqTypes>> + Send + Unpin + 'static,
+        K1: Sink<Url, Error = SendError> + Send + Unpin + 'static,
+        K2: Sink<Leaf<SeqTypes>, Error = SendError> + Send + Unpin + 'static,
+    {
+        let task_handle = async_std::task::spawn(Self::process_messages(
+            event_stream,
+            url_sender,
+            leaf_sender,
+        ));
+
+        Self {
+            task_handle: Some(task_handle),
+        }
+    }
+
+    /// [process_messages] is a function that will process messages from the
+    /// provided event stream.
+    async fn process_messages<S, K1, K2>(event_receiver: S, url_sender: K1, leaf_sender: K2)
+    where
+        S: Stream<Item = Event<SeqTypes>> + Send + Unpin + 'static,
+        K1: Sink<Url, Error = SendError> + Unpin,
+        K2: Sink<Leaf<SeqTypes>, Error = SendError> + Unpin,
+    {
+        let mut event_stream = event_receiver;
+        let mut url_sender = url_sender;
+        let mut leaf_sender = leaf_sender;
+        loop {
+            let event_result = event_stream.next().await;
+            let event = match event_result {
+                Some(event) => event,
+                None => {
+                    tracing::info!("event stream closed");
+                    break;
+                }
+            };
+
+            let Event { event, .. } = event;
+
+            match event {
+                EventType::Decide { leaf_chain, .. } => {
+                    for leaf_info in leaf_chain.iter() {
+                        let leaf = leaf_info.leaf.clone();
+
+                        let send_result = leaf_sender.send(leaf).await;
+                        if let Err(err) = send_result {
+                            tracing::info!("leaf sender closed: {}", err);
+                            return;
```

Comment:
This is suggested because we can potentially still make progress with the `url_sender`?  I think that could make sense, as we still might be able to make progress on `url_sender`.  The intention behind `HotShotEventProcessingTask` is to provide a single Task that can produce `Url`s and `Leaf`s without having to go through other sources.  So if the `leaf_receiver` is closed, then really we cannot provide any more leaf updates, and the state would just stagnate.

It seems that the only thing to do in this case would be exit this loop, and potentially attempt to restart it, either that or the program would have to stop.  Do you have any other ideas that would make more sense here?

Reply:
Yeah, you're right. It is actually better to kill the whole thing than continue in a bad state. I think we should probably just panic here and let the container orchestrator restart us, hopefully getting back into a good state. At the very least this should be an error level log

Reply:
Addressed in c0206d153d4d7ed2bbf676954ca6758123e24085.

I also added panics in the other tasks that are considered vital.